physionet-django/project/projectfiles/gcs.py

```diff
@@ -0,0 +1,146 @@
+import os
+
+from google.cloud.exceptions import NotFound, Conflict
+
+from physionet.gcs import GCSObject, GCSObjectException, create_bucket
+from project.projectfiles.base import BaseProjectFiles
+from project.utility import readable_size, FileInfo, DirectoryInfo
+
+
+class GCSProjectFiles(BaseProjectFiles):
+    def mkdir(self, path):
+        path = self._dir_path(path)
+
+        try:
+            GCSObject(path).mkdir()
+        except GCSObjectException:
+            raise FileExistsError
+
+    def rm(self, path):
+        try:
+            GCSObject(path).rm()
+        except NotFound:
+            GCSObject(self._dir_path(path)).rm()
+
+    def rm_dir(self, path, remove_zip=None):
+        path = self._dir_path(path)
+        GCSObject(path).rm()
+
+    def fwrite(self, path, content):
+        gcs_object = GCSObject(path)
+        if gcs_object.exists():
+            raise FileExistsError
+
+        gcs_object.upload_from_string(content)
+
+    def fput(self, path, file):
+        gcs_object = GCSObject(os.path.join(path, file.name))
+        if gcs_object.exists():
+            raise FileExistsError
+
+        gcs_object.upload_from_file(file)
+
+    def rename(self, source_path, target_path):
+        gcs_obj = GCSObject(source_path)
+        if not gcs_obj.exists():
+            gcs_obj = GCSObject(self._dir_path(source_path))
+            target_path = self._dir_path(target_path)
+
+        gcs_obj.rename(GCSObject(target_path))
+
+    def mv(self, source_path, target_path):
```

Comment:
Will that work if source_path points to an object and not a directory?

Reply:
I remember that when I implemented this originally, moving files and directories around (in the project edit pane) was quite tricky

Reply:
fixed